### PR TITLE
catalog: Read LD before catalog open

### PR DESCRIPTION
**Pull request by jkosh44**
This commit moves the sync with LaunchDarkly on startup from after opening the durable catalog to before opening the durable catalog. This allows us to use LaunchDarkly flags to determine how to open the durable catalog. Specifically this allows brand-new environments to use the catalog implementation specified in LaunchDarkly, without needing to reboot.

Works towards resolving #24218

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/environmentd/src/lib.rs [hunk 1]
```diff
@@ -362,22 +363,68 @@ impl Listeners {
...
+        } else {
+            None
+        };
+        let remote_system_parameters = load_remote_system_parameters(
+            &mut openable_adapter_storage,
+            system_parameter_sync_config.clone(),
+        )
+        .await?;
+
+        let catalog_kind_impl_ld = remote_system_parameters
```
Comment (danhhz) [inline src/environmentd/src/lib.rs:393]: Can we pull this out into a helper method, plz?

**Review by danhhz — APPROVED** [review]
LGTM!

File: src/environmentd/src/lib.rs (hunk 1 above)
Comment (jkosh44) [reply src/environmentd/src/lib.rs]: Done

**Comment by aalexandrov** [comment]
@jkosh44 will this change force an LD sync on every `environmend` boot or continue to do so only for brand new environments (which is the current behavior)? IIRC @benesch was keen to avoid making LD a hard dependency for an `environmentd` bootstrap when we were originally designing the feature.

**Comment by jkosh44** [comment]
> @jkosh44 will this change force an LD sync on every environmend boot or continue to do so only for brand new environments (which is the current behavior)?

No, we still only sync during startup for brand new environments. The only change is that we sync earlier in the startup process.

**Pull request merged**